cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -127,6 +124,104 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
 	return etcdClient, nil
 }
 
+// retrieveEtcdEndpointsFromClusterWithRetryOrMigrate returns the list of etcd endpoints,
+// by first trying to compose this information as reported by retrieveEtcdEndpointsFromClusterWithRetry.
+// If this list is empty, chances are that we haven't yet migrated this information to reside
+// on pod annotations, fallback in that case to retrieveEtcdEndpointsFromClusterStatusWithRetry,
+// that will compose this information out of the ClusterStatus struct present in the kubeadm-config ConfigMap.
+//
+// TODO: remove after 1.20, when the ClusterStatus struct is removed from the kubeadm-config ConfigMap;
+//       callers of this function should call retrieveEtcdEndpointsFromClusterWithRetry directly when that
+//       happens.
+func retrieveEtcdEndpointsFromClusterWithRetryOrMigrate(client clientset.Interface) ([]string, error) {
+	etcdEndpoints, err := retrieveEtcdEndpointsFromClusterWithRetry(client)
+	if err != nil {
+		return retrieveEtcdEndpointsFromClusterStatusWithRetry(client)
+	}
+	return etcdEndpoints, nil
+}
+
+// retrieveEtcdEndpointsFromClusterWithRetry returns the list of endpoints as reported on etcd's
+// pods labels. It retries the operation if the list of endpoints is empty (the apiserver still
+// doesn't know about any etcd pod -- could still being populated), or if any unexpected error
+// happened when listing pods.
+func retrieveEtcdEndpointsFromClusterWithRetry(client clientset.Interface) ([]string, error) {
+	etcdEndpoints := []string{}
+	var lastErr error
+	err := wait.ExponentialBackoff(clientsetretry.DefaultRetry, func() (bool, error) {
+		if etcdEndpoints, lastErr = retrieveEtcdEndpointsFromCluster(client); lastErr != nil {
+			return false, nil
+		}
+		// If the list of etcd endpoints is empty we want to retry: this can happen if joining a secondary
+		// control plane while the primary control plane didn't mirror its static pods yet.
+		return len(etcdEndpoints) > 0, nil
+	})
+	if err != nil {
+		if lastErr != nil {
+			return []string{}, errors.Wrap(lastErr, "could not retrieve the list of etcd endpoints")
+		}
+		return []string{}, errors.Wrap(err, "could not retrieve the list of etcd endpoints")
+	}
+	return etcdEndpoints, nil
+}
+
+// retrieveEtcdEndpointsFromCluster returns the list of endpoints as reported on etcd's pods labels.
+func retrieveEtcdEndpointsFromCluster(client clientset.Interface) ([]string, error) {
```

Comment:
The name sounds confusing. Better clarify it as `FromPod` than just `FromCluster`.